akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,237 @@
+
+      override def onPush(): Unit = {
+        try {
+          val elem = grab(in)
+          val holder = new Holder[Partition, In, Out](
+            incoming = elem,
+            outgoing = NotYetThere,
+            partition = partitioner(elem),
+            cb = futureCB)
+
+          buffer.enqueue(holder)
+        } catch {
+          case NonFatal(ex) =>
+            if (decider(ex) == Supervision.Stop) failStage(ex)
+        }
+
+        pullIfNeeded()
+      }
+
+      override def onUpstreamFinish(): Unit = if (buffer.isEmpty) completeStage()
+
+      @tailrec
+      private def dropCompletedThenPushIfPossible(partition: Partition): Unit = {
+        val holderOpt = buffer.peekPartition(partition)
+        holderOpt match {
+          case None => pushNextIfPossible()
+          case Some(holder) =>
+            holder.outgoing match {
+              case NotYetThere                                                                              => pushNextIfPossible()
+              case Failure(NonFatal(ex)) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop =>
+                // Could happen if this finds the failed future before the async callback runs
+                failStage(ex)
+
+              case Failure(NonFatal(_)) | Success(_) =>
+                buffer.dropOnlyPartitionHead(partition)
+                dropCompletedThenPushIfPossible(partition)
+
+              case Failure(ex) =>
+                // fatal exception in the buffer, not sure if this can actually happen, but for completeness...
+                throw ex
+            }
+        }
+      }
+
+      @tailrec
+      private def pushNextIfPossible(): Unit =
+        if (buffer.isEmpty) pullIfNeeded()
+        else if (buffer.peek().outgoing eq NotYetThere) pullIfNeeded()
+        else if (isAvailable(out)) {
+          // We peek instead of dequeue so that we can push out an element before
+          //  removing from the queue (since a dequeue or dropHead can cause re-entry)
+          val holder = buffer.peek()
+          holder.outgoing match {
+            case Success(elem) =>
+              if (elem != null) {
+                push(out, elem)
+                buffer.dropHead()
+                pullIfNeeded()
+              } else {
+                // elem is null
+                buffer.dropHead()
+                pullIfNeeded()
+                pushNextIfPossible()
+              }
+
+            case Failure(NonFatal(ex)) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop =>
+              failStage(ex)
+
+            case Failure(NonFatal(_)) =>
+              buffer.dropHead()
+              pushNextIfPossible() // try the next element
+
+            case Failure(ex) => throw ex
+          }
+        }
+
+      private def pullIfNeeded(): Unit = {
+        if (isClosed(in) && buffer.isEmpty) completeStage()
+        else if (buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)
```

Comment:
We keep pulling as long as there are allowed concurrent jobs, but they may also be limited by the `perPartition` limit. If one partition gets a burst of `parallelism` elements and has a `perPartition` limit (which is the point) and is slower it will block the other partitions from continuing until it has worked through one (or all?) elements.

Would be nice if we wouldn't block the other partitions like that. Not. sure how it would work, but for `groupBy` we don't backpressure until one downstream/partition is backpressuring and we got an element for that backpressured partition.

Reply:
Adding an explicit test case for that behavior (`"not backpressure based on perPartition limit"`)...